File: src/components/menu.ts
/**
 * Interactive nested inline menu system with dynamic navigation and callback routing.
 *
 * @packageDocumentation
 */

import type { InlineKeyboardMarkup, InlineKeyboardButton } from "../client/types.js";
import type { CallbackContext, MenuContextControl } from "../kernel/context.js";
import type { MiddlewareFn } from "../kernel/dispatcher.js";

/**
 * Dynamic label evaluator function for menu buttons.
 */
export type MenuLabel = string | ((ctx: CallbackContext) => string | Promise<string>);

/**
 * Click handler callback invoked when a menu button is pressed.
 */
export type MenuButtonHandler = (ctx: CallbackContext) => Promise<void> | void;

/**
 * Navigation hook callback invoked during menu transitions.
 */
export type MenuNavigationHandler = (ctx: CallbackContext) => Promise<void> | void;

/**
 * Internal representation of a menu button item.
 */
type MenuButtonItem =
  | {
      type: "text";
      label: MenuLabel;
      handler: MenuButtonHandler;
    }
  | {
      type: "submenu";
      label: MenuLabel;
      targetMenu: Menu;
      onNavigate?: MenuNavigationHandler;
    }
  | {
      type: "back";
      label: MenuLabel;
      onNavigate?: MenuNavigationHandler;
    }
  | {
      type: "url";
      label: MenuLabel;
      url: string;
    };

/**
 * Interactive, composable inline keyboard menu component.
 *
 * Supports multi-level nested menus, dynamic labels, button click handlers,
 * in-place message navigation, and automatic middleware routing.
 *
 * @example
 * ```ts
 * const mainMenu = new Menu("main")
 *   .text("Profile", async (ctx) => {
 *     await ctx.reply("Opening profile...");
 *   })
 *   .row();
 *
 * const settingsMenu = new Menu("settings")
 *   .text("Toggle Dark Mode", async (ctx) => {
 *     await ctx.answerCallbackQuery({ text: "Dark mode toggled!" });
 *   })
 *   .row()
 *   .back();
 *
 * mainMenu.submenu("Settings", settingsMenu);
 *
 * app.use(mainMenu);
 * ```
 */
export class Menu {
  /**
   * Unique string identifier for this menu.
   */
  public readonly id: string;

  /**
   * Reference to the parent {@link Menu} if this instance is a submenu.
   */
  public parent?: Menu;

  /**
   * Internal 2D grid of button rows.
   */
  private readonly rows: MenuButtonItem[][] = [[]];

  /**
   * Registered child submenus indexed by their menu ID.
   */
  private readonly submenus = new Map<string, Menu>();

  /**
   * Constructs a new {@link Menu} instance.
   *
   * @param id - Unique string identifier for the menu.
   */
  constructor(id: string) {
    if (!id || id.trim().length === 0) {
      throw new TypeError("Menu id must be a non-empty string.");
    }
    this.id = id.trim();
  }

  /**
   * Appends an interactive callback button with a click handler to the current row.
   *
   * @param label - Static string label or dynamic label function.
   * @param handler - Callback function executed when the button is clicked.
   * @returns This {@link Menu} instance for chaining.
   *
   * @example
   * ```ts
   * menu.text("Click Me", async (ctx) => {
   *   await ctx.reply("Button clicked!");
   * });
   * ```
   */
  public text(label: MenuLabel, handler: MenuButtonHandler): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({
        type: "text",
        label,
        handler,
      });
    }
    return this;
  }

  /**
   * Advances the menu builder to start a new row of buttons.
   *
   * @returns This {@link Menu} instance for chaining.
   */
  public row(): this {
    if (
      this.rows.length === 0 ||
      (this.rows[this.rows.length - 1] && this.rows[this.rows.length - 1]!.length > 0)
    ) {
      this.rows.push([]);
    }
    return this;
  }

  /**
   * Appends a navigation button leading to a child {@link Menu}.
   *
   * @param label - Static string label or dynamic label function.
   * @param targetMenu - The destination submenu to open.
   * @param onNavigate - Optional callback hook executed before switching menus.
   * @returns This {@link Menu} instance for chaining.
   *
   * @example
   * ```ts
   * menu.submenu("Preferences", preferencesMenu);
   * ```
   */
  public submenu(label: MenuLabel, targetMenu: Menu, onNavigate?: MenuNavigationHandler): this {
    targetMenu.parent = this;
    this.submenus.set(targetMenu.id, targetMenu);

    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({
        type: "submenu",
        label,
        targetMenu,
        onNavigate,
      });
    }
    return this;
  }

  /**
   * Appends a back button that returns to the parent menu.
   *
   * @param label - Button label text (default: `"Back"`).
   * @param onNavigate - Optional callback hook executed before navigating back.
   * @returns This {@link Menu} instance for chaining.
   *
   * @example
   * ```ts
   * subMenu.back("Go Back");
   * ```
   */
  public back(label: MenuLabel = "Back", onNavigate?: MenuNavigationHandler): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({
        type: "back",
        label,
        onNavigate,
      });
    }
    return this;
  }

  /**
   * Appends an external URL button to the current row.
   *
   * @param label - Static string label or dynamic label function.
   * @param url - HTTP or tg:// URL opened when the button is clicked.
   * @returns This {@link Menu} instance for chaining.
   */
  public url(label: MenuLabel, url: string): this {
    const currentRow = this.rows[this.rows.length - 1];
    if (currentRow) {
      currentRow.push({
        type: "url",
        label,
        url,
      });
    }
    return this;
  }

  /**
   * Synchronously builds the {@link InlineKeyboardMarkup} for this menu.
   *
   * @param ctx - Optional {@link CallbackContext} for dynamic label resolution.
   * @returns Constructed inline keyboard markup.
   *
   * @example
   * ```ts
   * await ctx.reply("Choose an option:", {
   *   reply_markup: menu.build(ctx),
   * });
   * ```
   */
  public build(ctx?: CallbackContext): InlineKeyboardMarkup {
    const inline_keyboard: InlineKeyboardButton[][] = [];

    for (let r = 0; r < this.rows.length; r++) {
      const row = this.rows[r];
      if (!row || row.length === 0) continue;

      const keyboardRow: InlineKeyboardButton[] = [];
      for (let c = 0; c < row.length; c++) {
        const item = row[c]!;
        let text: string;
        if (typeof item.label === "function") {
          try {
            const res = item.label(ctx as CallbackContext);
            text = typeof res === "string" ? res : "";
          } catch {
            text = "";
          }
        } else {
          text = item.label;
        }

        switch (item.type) {
          case "text":
            keyboardRow.push({
              text,
              callback_data: `m:${this.id}:b:${r}:${c}`,
            });
            break;
          case "submenu":
            keyboardRow.push({
              text,
              callback_data: `m:${this.id}:s:${item.targetMenu.id}:${r}:${c}`,
            });
            break;
          case "back":
            keyboardRow.push({
              text,
              callback_data: `m:${this.id}:k:${r}:${c}`,
            });
            break;
          case "url":
            keyboardRow.push({
              text,
              url: item.url,
            });
            break;
        }
      }

      if (keyboardRow.length > 0) {
        inline_keyboard.push(keyboardRow);
      }
    }

    return { inline_keyboard };
  }

  /**
   * Asynchronously renders the {@link InlineKeyboardMarkup}, resolving any async dynamic labels.
   *
   * @param ctx - Optional {@link CallbackContext} for dynamic label evaluation.
   * @returns Resolves with the constructed {@link InlineKeyboardMarkup}.
   */
  public async render(ctx?: CallbackContext): Promise<InlineKeyboardMarkup> {
    const inline_keyboard: InlineKeyboardButton[][] = [];

    for (let r = 0; r < this.rows.length; r++) {
      const row = this.rows[r];
      if (!row || row.length === 0) continue;

      const keyboardRow: InlineKeyboardButton[] = [];
      for (let c = 0; c < row.length; c++) {
        const item = row[c]!;
        let text: string;
        if (typeof item.label === "function") {
          try {
            const res = await item.label(ctx as CallbackContext);
            text = typeof res === "string" ? res : "";
          } catch {
            text = "";
          }
        } else {
          text = item.label;
        }

        switch (item.type) {
          case "text":
            keyboardRow.push({
              text,
              callback_data: `m:${this.id}:b:${r}:${c}`,
            });
            break;
          case "submenu":
            keyboardRow.push({
              text,
              callback_data: `m:${this.id}:s:${item.targetMenu.id}:${r}:${c}`,
            });
            break;
          case "back":
            keyboardRow.push({
              text,
              callback_data: `m:${this.id}:k:${r}:${c}`,
            });
            break;
          case "url":
            keyboardRow.push({
              text,
              url: item.url,
            });
            break;
        }
      }

      if (keyboardRow.length > 0) {
        inline_keyboard.push(keyboardRow);
      }
    }

    return { inline_keyboard };
  }

  /**
   * Finds a menu instance by its identifier anywhere in this menu's hierarchy.
   *
   * @param menuId - Identifier of the menu to locate.
   * @param visited - Internal set of visited menus to avoid cycles.
   * @returns The matching {@link Menu} instance or `undefined`.
   */
  public findMenu(menuId: string, visited: Set<Menu> = new Set()): Menu | undefined {
    if (visited.has(this)) return undefined;
    visited.add(this);

    if (this.id === menuId) return this;

    for (const sub of this.submenus.values()) {
      const found = sub.findMenu(menuId, visited);
      if (found) return found;
    }

    if (this.parent) {
      return this.parent.findMenu(menuId, visited);
    }

    return undefined;
  }

  /**
   * Creates a {@link MiddlewareFn} that automatically intercepts matching callback queries,
   * dispatches button handlers, and performs in-place menu navigation.
   *
   * @returns A middleware function for use with `app.use()`.
   */
  public middleware(): MiddlewareFn {
    return async (ctx: CallbackContext, next: () => Promise<void>): Promise<void> => {
      const data = ctx.update?.callback_query?.data;
      if (!data || !data.startsWith("m:")) {
        return next();
      }

      const parts = data.split(":");
      const targetMenuId = parts[1];
      const action = parts[2];

      if (!targetMenuId || !action) {
        return next();
      }

      const menu = this.findMenu(targetMenuId);
      if (!menu) {
        return next();
      }

      const rowIndex = Number(parts[parts.length - 2]);
      const colIndex = Number(parts[parts.length - 1]);
      const buttonItem = menu.rows[rowIndex]?.[colIndex];

      if (!buttonItem) {
        return next();
      }

      const chatId = ctx.update?.effective_chat?.id;
      const messageId = ctx.update?.effective_message?.message_id;

      const menuControl: MenuContextControl = {
        update: async (): Promise<void> => {
          if (chatId && messageId) {
            const updatedMarkup = await menu.render(ctx);
            try {
              await ctx.bot.editMessageReplyMarkup({
                chat_id: chatId,
                message_id: messageId,
                reply_markup: updatedMarkup,
              });
            } catch {
              // Ignore unchanged markup error
            }
          }
        },
        nav: async (targetMenu: unknown): Promise<void> => {
          if (chatId && messageId && targetMenu instanceof Menu) {
            const nextMarkup = await targetMenu.render(ctx);
            try {
              await ctx.bot.editMessageReplyMarkup({
                chat_id: chatId,
                message_id: messageId,
                reply_markup: nextMarkup,
              });
            } catch {
              // Ignore
            }
          }
        },
        back: async (): Promise<void> => {
          if (menu.parent && chatId && messageId) {
            const parentMarkup = await menu.parent.render(ctx);
            try {
              await ctx.bot.editMessageReplyMarkup({
                chat_id: chatId,
                message_id: messageId,
                reply_markup: parentMarkup,
              });
            } catch {
              // Ignore
            }
          }
        },
      };

      ctx.menu = menuControl;

      let isCallbackAnswered = false;
      const originalAnswer = ctx.answerCallbackQuery.bind(ctx);
      ctx.answerCallbackQuery = async (options) => {
        isCallbackAnswered = true;
        return originalAnswer(options);
      };

      if (buttonItem.type === "text" && action === "b") {
        await buttonItem.handler(ctx);
        if (!isCallbackAnswered) {
          try {
            await ctx.answerCallbackQuery();
          } catch {
            // Ignore
          }
        }
        return;
      }

      if (buttonItem.type === "submenu" && action === "s") {
        if (buttonItem.onNavigate) {
          await buttonItem.onNavigate(ctx);
        }

        if (chatId && messageId) {
          const nextMarkup = await buttonItem.targetMenu.render(ctx);
          try {
            await ctx.bot.editMessageReplyMarkup({
              chat_id: chatId,
              message_id: messageId,
              reply_markup: nextMarkup,
            });
          } catch {
            // Ignore
          }
        }

        if (!isCallbackAnswered) {
          try {
            await ctx.answerCallbackQuery();
          } catch {
            // Ignore
          }
        }
        return;
      }

      if (buttonItem.type === "back" && action === "k") {
        if (menu.parent) {
          if (buttonItem.onNavigate) {
            await buttonItem.onNavigate(ctx);
          }

          if (chatId && messageId) {
            const parentMarkup = await menu.parent.render(ctx);
            try {
              await ctx.bot.editMessageReplyMarkup({
                chat_id: chatId,
                message_id: messageId,
                reply_markup: parentMarkup,
              });
            } catch {
              // Ignore
            }
          }

          if (!isCallbackAnswered) {
            try {
              await ctx.answerCallbackQuery();
            } catch {
              // Ignore
            }
          }
        }
        return;
      }

      return next();
    };
  }
}

File: packages/python/src/telebot_py/components/menu/menu.py
"""Interactive nested inline menu builder (parity with node menu/menu.ts).

Callback routing uses the node scheme so buttons are wire-compatible:
``m:{menu}:b:{row}:{col}`` (text), ``m:{menu}:s:{target}:{row}:{col}``
(submenu), ``m:{menu}:k:{row}:{col}`` (back). The dispatch middleware from
the node version lands with the kernel integration pass; this module covers
the builder/render/lookup surface.
"""

from __future__ import annotations

import typing as t

from telebot_py.components.menu.renderer import build_menu_keyboard, render_menu_keyboard
from telebot_py.components.menu.types import (
    BackButton,
    MenuButtonHandler,
    MenuButtonItem,
    MenuLabel,
    MenuNavigationHandler,
    SubmenuButton,
    TextButton,
    UrlButton,
)
from telebot_py.types import InlineKeyboardMarkup

if t.TYPE_CHECKING:
    from telebot_py.kernel.context import CallbackContext

_DEFAULT_BACK_LABEL = "Back"


class Menu:
    """A nested inline menu with dynamic navigation and callback routing.

    Example:
        >>> settings = Menu("settings").back("Go back")
        >>> main = (
        ...     Menu("main")
        ...     .text("Click me", on_click)
        ...     .row()
        ...     .submenu("Settings", settings)
        ... )
        >>> await bot.send_message(chat_id, "Menu:", reply_markup=main.build())

    Attributes:
        id: Unique string identifier for this menu.
        parent: The parent menu when this instance is a submenu.
    """

    def __init__(self, id: str) -> None:  # noqa: A002 - parity with the node API
        """Create a new menu.

        Args:
            id: Unique identifier; embedded in every button's callback data.

        Raises:
            ValueError: If ``id`` is empty or whitespace-only.
        """
        if not id or not id.strip():
            msg = "Menu id must be a non-empty string."
            raise ValueError(msg)
        self.id = id.strip()
        self.parent: Menu | None = None
        self._rows: list[list[MenuButtonItem]] = [[]]
        self._submenus: dict[str, Menu] = {}

    def text(self, label: MenuLabel, handler: MenuButtonHandler) -> Menu:
        """Append an interactive callback button to the current row.

        Args:
            label: Static label or (possibly async) resolver.
            handler: Callback executed when the button is clicked.

        Returns:
            This menu, for chaining.

        Example:
            >>> menu.text("Click me", on_click)
        """
        self._rows[-1].append(TextButton(label=label, handler=handler))
        return self

    def row(self) -> Menu:
        """Advance to a new button row; no-op if the current row is empty.

        Returns:
            This menu, for chaining.
        """
        if self._rows[-1]:
            self._rows.append([])
        return self

    def submenu(
        self,
        label: MenuLabel,
        target_menu: Menu,
        on_navigate: MenuNavigationHandler | None = None,
    ) -> Menu:
        """Append a navigation button leading to a child menu.

        Args:
            label: Static label or (possibly async) resolver.
            target_menu: The destination submenu to open.
            on_navigate: Optional hook executed before switching menus.

        Returns:
            This menu, for chaining.

        Example:
            >>> menu.submenu("Preferences", preferences_menu)
        """
        target_menu.parent = self
        self._submenus[target_menu.id] = target_menu
        self._rows[-1].append(
            SubmenuButton(label=label, target_menu=target_menu, on_navigate=on_navigate)
        )
        return self

    def back(
        self,
        label: MenuLabel = _DEFAULT_BACK_LABEL,
        on_navigate: MenuNavigationHandler | None = None,
    ) -> Menu:
        """Append a back button returning to the parent menu.

        Args:
            label: Button label; defaults to 'Back'.
            on_navigate: Optional hook executed before navigating back.

        Returns:
            This menu, for chaining.

        Example:
            >>> submenu.back("Go back")
        """
        self._rows[-1].append(BackButton(label=label, on_navigate=on_navigate))
        return self

    def url(self, label: MenuLabel, url: str) -> Menu:
        """Append an external URL button to the current row.

        Args:
            label: Static label or (possibly async) resolver.
            url: HTTP or tg:// URL opened when the button is clicked.

        Returns:
            This menu, for chaining.
        """
        self._rows[-1].append(UrlButton(label=label, url=url))
        return self

    def build(self, ctx: CallbackContext | None = None) -> InlineKeyboardMarkup:
        """Synchronously build this menu's inline keyboard markup.

        Args:
            ctx: Optional context for dynamic label resolution.

        Returns:
            The constructed InlineKeyboardMarkup.

        Example:
            >>> await bot.send_message(chat_id, "Choose:", reply_markup=menu.build(ctx))
        """
        return build_menu_keyboard(self.id, self._rows, ctx)

    async def render(self, ctx: CallbackContext | None = None) -> InlineKeyboardMarkup:
        """Asynchronously build the markup, awaiting async label resolvers.

        Args:
            ctx: Optional context for dynamic label resolution.

        Returns:
            The constructed InlineKeyboardMarkup.
        """
        return await render_menu_keyboard(self.id, self._rows, ctx)

    def find_menu(self, menu_id: str, visited: set[Menu] | None = None) -> Menu | None:
        """Find a menu by identifier anywhere in this menu's hierarchy.

        Args:
            menu_id: Identifier of the menu to locate.
            visited: Internal cycle guard; omit on the first call.

        Returns:
            The matching menu, or None when not found.
        """
        seen = visited if visited is not None else set()
        if self in seen:
            return None
        seen.add(self)

        if self.id == menu_id:
            return self
        for submenu in self._submenus.values():
            found = submenu.find_menu(menu_id, seen)
            if found is not None:
                return found
        if self.parent is not None:
            return self.parent.find_menu(menu_id, seen)
        return None
